File: src/main/java/com/zepben/vertxutils/json/JsonValueExtractors.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.json;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import javax.annotation.Nullable;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;

@EverythingIsNonnullByDefault
@SuppressWarnings("WeakerAccess")
public class JsonValueExtractors {

    @Nullable
    public static Double getDouble(JsonObject jsonObject, String key) {
        try {
            return jsonObject.getDouble(key);
        } catch (ClassCastException e) {
            return checkForNaN(Objects.requireNonNull(getStringStrict(jsonObject, key)), e);
        }
    }

    @Nullable
    public static Double getDouble(JsonArray jsonArray, int index) {
        try {
            return jsonArray.getDouble(index);
        } catch (ClassCastException e) {
            return checkForNaN(Objects.requireNonNull(getStringStrict(jsonArray, index)), e);
        }
    }

    @Nullable
    public static Path getPath(JsonObject json, String key) {
        String string = getStringStrict(json, key);
        if (string != null)
            return Paths.get(string);
        else
            return null;
    }

    /**
     * JsonObject::getString in VertX automatically converts some non-strings to strings, such as numbers.
     * This function ensures that the value in the object is actually a string.
     * @return jsonObject[key] if it's a string, null if it isn't found
     * @throws ClassCastException if jsonObject[key] is not a string
     */
    @Nullable
    public static String getStringStrict(JsonObject jsonObject, String key) throws ClassCastException {
        return (String) jsonObject.getValue(key);
    }

    /**
     * JsonObject::getString in VertX automatically converts some non-strings to strings, such as numbers.
     * This function ensures that the value in the array is actually a string.
     * @return jsonArray[index] if it's a string, null if it isn't found
     * @throws ClassCastException if jsonArray[index] is not a string
     */
    @Nullable
    public static String getStringStrict(JsonArray jsonArray, int index) throws ClassCastException {
        return (String) jsonArray.getValue(index);
    }

    private static Double checkForNaN(String value, ClassCastException e) {
        if (value.equals("NaN"))
            return Double.NaN;
        else
            throw e;
    }
}

File: src/main/java/com/zepben/vertxutils/json/JsonValueExtractors.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.json

import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import java.nio.file.Path
import java.nio.file.Paths

object JsonValueExtractors {

    /**
     * A wrapper for the built in `getDouble` that also supports "NaN" strings.
     */
    fun getDouble(jsonObject: JsonObject, key: String): Double? =
        try {
            jsonObject.getDouble(key)
        } catch (e: ClassCastException) {
            checkForNaN(requireNotNull(getStringStrict(jsonObject, key)), e)
        }

    /**
     * A wrapper for the built in `getDouble` that also supports "NaN" strings.
     */
    fun getDouble(jsonArray: JsonArray, pos: Int): Double? =
        try {
            jsonArray.getDouble(pos)
        } catch (e: ClassCastException) {
            checkForNaN(requireNotNull(getStringStrict(jsonArray, pos)), e)
        }

    fun getPath(json: JsonObject, key: String): Path? =
        getStringStrict(json, key)?.let { Paths.get(it) }

    /**
     * JsonObject::getString in VertX automatically converts some non-strings to strings, such as numbers.
     * This function ensures that the value in the object is actually a string.
     * @return jsonObject[key] if it's a string, null if it isn't found
     * @throws ClassCastException if jsonObject[key] is not a string
     */
    @Throws(ClassCastException::class)
    fun getStringStrict(jsonObject: JsonObject, key: String): String? =
        jsonObject.getValue(key) as String?

    /**
     * JsonObject::getString in VertX automatically converts some non-strings to strings, such as numbers.
     * This function ensures that the value in the array is actually a string.
     * @return jsonArray[pos] if it's a string, null if it isn't found
     * @throws ClassCastException if jsonArray[pos] is not a string
     */
    @Throws(ClassCastException::class)
    fun getStringStrict(jsonArray: JsonArray, pos: Int): String? =
        jsonArray.getValue(pos) as String?

    private fun checkForNaN(value: String, e: ClassCastException): Double =
        when (value) {
            "NaN" -> Double.NaN
            else -> throw e
        }

}
